### PR TITLE
[REG-1848] Address instrumentation conflicts and improve error handling

### PR DESCRIPTION
**Pull request by svolokh**
This is a second pass at addressing issues that can occur with the input instrumentation:
1. Check the `EditorApplication.isCompiling` and `EditorApplication.isUpdating` flags. If either is true, defer the instrumentation until later.
2. Remove the inner exception handler in InstrumentAssemblyIfNeeded (we already have exception handlers in the callers), and broaden the exceptions handled to all Exceptions (this will handle any errors that could come out of Mono.Cecil's own APIs).

I've tested this by re-opening Boss Room and some other Unity projects from scratch (deleting the Library folder). The whole instrumentation process works in my testing.

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/ActionManager/Actions/MouseHoverObjectAction.cs [hunk 1]
```diff
@@ -1,8 +1,6 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Text;
-using System.Collections.Generic;
```
Comment (svolokh) [inline src/gg.regression.unity.bots/Runtime/Scripts/ActionManager/Actions/MouseHoverObjectAction.cs:4]: Fixing a warning that was introduced from an earlier PR

**Comment by RG-nAmKcAz** [comment]
@svolokh This looks like it will prevent us from starting while Unity is actively compiling.

Do we have a way to either interrupt us, or block Unity from compiling until we finish if a user makes & saves a script change or attaches the debugger while we're mid compilation ?

**Comment by svolokh** [comment]
@RG-nAmKcAz The logic for deferring while Unity is compiling is only applied to the initial instrumentation pass (when the RG SDK is first loaded and it iterates through the assemblies to check if any need instrumentation). This is generally where the source of IO errors comes from.

If a **new** compilation is triggered while our SDK is active, then this will instead trigger the `OnAssemblyCompiled` callback and our instrumentation will still be applied during Unity's compilation pipeline. So this will still work correctly if the user makes changes.

As far as locking the Unity editor from compiling, the best thing I can find is [EditorApplication.LockReloadAssemblies()](https://docs.unity3d.com/ScriptReference/EditorApplication.LockReloadAssemblies.html), but it's not clear to me whether this will be an improvement.

**Review by RG-nAmKcAz — APPROVED** [review]
I think we may still need 1 more pass to handle the discussion in this PR about blocking unity from recompiling changes while we are in process, but this is another good step.

**Pull request merged**